core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+          session.action,
+          params,
+          Some(session.action.limits.timeout.duration + 1.minute), // wait for result
+          Some(session.activationId)) // cause is session id
+
+      response.onComplete {
+        case Failure(t) =>
+          // invocation failure
+          val response = ActivationResponse.whiskError(compositionActivationFailure)
+          completeAppActivation(user, session, response)
+        case Success(Left(activationId)) =>
+          // invocation timeout
+          session.logs += activationId
+          val response = ActivationResponse.whiskError(compositionActivationTimeout(activationId))
+          completeAppActivation(user, session, response)
+        case Success(Right(activation)) =>
+          // successful invocation
+          session.logs += activation.activationId
+          session.duration += activation.duration.getOrElse(activation.end.toEpochMilli - activation.start.toEpochMilli)
+
+          val result = activation.resultAsJson
+
+          // extract params from result
+          val params = result.getFields(WhiskActivation.paramsField).headOption.map { p =>
+            Try(p.asJsObject).getOrElse(JsObject(WhiskActivation.valueField -> p)) // ensure params is a dictionary
+          }
+
+          // update session state
+          session.state = result.getFields(WhiskActivation.stateField).headOption.flatMap { p =>
+            Try(Some(p.asJsObject)).getOrElse(None)
+          }
+
+          // extract next action from result and invoke
+          result.getFields(WhiskActivation.actionField).headOption match {
+            case None =>
+              // no next action, end composition execution, return to caller
+              val response = ActivationResponse(activation.response.statusCode, Some(params.getOrElse(result)))
+              completeAppActivation(user, session, response)
+            case Some(next) =>
+              Try(next.convertTo[EntityPath]) match {
+                case Failure(t) =>
+                  // parsing failure
+                  val response = ActivationResponse.applicationError(compositionComponentInvalid(next))
+                  completeAppActivation(user, session, response)
+                case Success(next) =>
+                  // resolve and invoke next action
+                  val fqn = (if (next.defaultPackage) EntityPath.DEFAULT.addPath(next) else next)
+                    .resolveNamespace(user.namespace)
+                    .toFullyQualifiedEntityName
+                  val resource = Resource(fqn.path, Collection(Collection.ACTIONS), Some(fqn.name.asString))
+                  entitlementProvider.check(user, Privilege.ACTIVATE, Set(resource), noThrottle = true).onComplete {
+                    case Failure(t) =>
+                      // failed entitlement check
+                      val response =
+                        ActivationResponse.applicationError(compositionComponentNotAccessible(next.toString))
```

Comment:
s/toString/asString/ which is defined on a fqn and preferred as the canonical string serializer.

Reply:
the entitlement check should fail with a RejectRequest exception which contains a reason for the failure. Did you consider use that instead of a new error message (which I'm ok with - just curious if you considered it).

Reply:
You mean doing this?

```javascript
case Failure(t: RejectRequest) => ActivationResponse(t.code.intValue, Some(t.toString.toJson))
```

Reply:
I considered this more - I think your error is adequate. We can reconsider in the future. I don't think we want to leak the `code` here since it's an HTTP Status code which doesn't make sense here. The message is an option, so you'd have to map on that and get either the custom message or the default message for the `code` otherwise. I think this doesn't make the error reporting any more meaningful.